File: util/deapi/c++/main/dealingtools/orc/deapi/MessageQueue.cc
/* ----------------------------------------------------------------------
   MessageQueue - deapi message queue implementation

   Original author: John McMullan (2004)

   $Author: john $
   $Date: 2004/05/10 14:06:02 $
   $Source: /usr/export/cvsroot/util/deapi/c++/main/dealingtools/orc/deapi/MessageQueue.cc,v $
   $Revision: 1.1.1.1 $

   ---------------------------------------------------------------------- */

static char rcsid[] = "$Id: MessageQueue.cc,v 1.1.1.1 2004/05/10 14:06:02 john Exp $";

#include <dt/util.h>

#include <pthread.h>

#include <string>
#include <hash_map>
#include <queue>

#include "Message.h"
#include "MessageQueue.h"
#include "Exception.h"
#include "Connection.h"

static const char *module = "MessageQueue";

namespace deapi {

  MessageQueue::MessageQueue(const std::string& hostname,
			     const std::string& service,
			     const std::string& password,
			     bool debug)
    : deapi(hostname, service) {
    Initialize();
    Login(loginid, password, debug);
  }

  MessageQueue::MessageQueue(const std::string& hostname,
			     const std::string& service,
			     const std::string& loginid,
			     const std::string& password)
    : deapi(hostname, service) {
    Initialize();
    Login(loginid, password, false);
  }

  void MessageQueue::Initialize() {
    pthread_t tid;
    pthread_attr_t attr;
    pthread_attr_init(&attr);

    pthread_mutex_init(&lock, 0);
    pthread_mutex_init(&goLock, 0);
    pthread_cond_init(&go, 0);

    // kick off our recieve thread
    if ((int ret =
	 pthread_create(&tid, &attr, MessageQueue::Run, this)) != 0) {
      switch (ret) {
      case EAGAIN:
	throw Exception(ERR_SYSERR, strerror(ret));
	break;
      case EINVAL:
	throw Exception(ERR_SYSERR, strerror(ret));
	break;
      }
    }
  }

  void MessageQueue::Login(const std::string& loginid,
			   const std::string& password,
			   bool debug) {
    std::string loginmsg = "{message_info={message_type=login}|login_id=";
    loginmsg += loginId + "|password=";
    loginmsg += password;
    if (debug) {
      loginmsg += "|debug=on";
    }
    loginmsg += "}";

    Send(loginmsg);
    Recieve("login");
  }

  void MessageQueue::Logout() {
    try {
      Send("{message_info={message_type=logout}}");
    } catch (...) {
      // I can't think why I need to worry...
    }
  }

  void MessageQueue::Run(void *closure) {
    MessageQueue *self = static_cast<MessageQueue *>(closure);
    dt::Log *log = dt::Log::Instance();

    while (true) {
      try {
	Message *reply = self->deapi.recieve();
	std::string id =
	  MessageQueue::MessageId(reply.Type(), reply.PrivateId());
      } catch (...) {
	log->Debug(module, "returning from Run() thread\n");
	pthread_exit();
      }
      
      pthread_mutex_lock(&lock);
      MessageCallback *callback = messageCallbacks.find(id);
      if (callback != NULL) {
	// there was a callback registered
	try {
	  if ((int err = reply->Error()) != 0) {
	    callback->OnError(err, reply->ErrorDescription());
	  } else {
	    callback->OnMessage(reply);
	  }
	} catch (...) {
	  log->Warning(module, "Unable to get error status from message\n");
	}
      } else {
	// no callback, stick it in the queue and give that a poke
	MsgQueue *msgQueue = messageQueues.find(id);
	if (msgQueue = NULL) {
	  msgQueue = new MsgQueue();
	  messageQueues[id] = msgQueue;
	}
	
	// poke
	msgQueue.push(reply);
	pthread_cond_signal(&go);
      }
      pthread_mutex_unlock(&goLock);
    }
  }

  void MessageQueue::RegisterCallback(const std::string& messageType,
				      const std::string& privateId,
				      MessageCallback *callback) {
    std::string id = MessageQueue::MessageId(messageType, privateId);
    pthread_mutex_lock(&lock);
    MessageCallback *old = messageCallbacks.find(id);
    if (old != 0)
      delete old;
    messageCallbacks[id] = callback;
    pthread_mutex_unlock(&lock);
  }
  void MessageQueue::RegisterCallback(std::string& messageType,
				      MessageCallback *callback) {
    RegisterCallback(messageType, "", callback);
  }

  void MessageQueue::RemoveCallback(std::string& messageType,
				    std::string& privateId) {
    std::string id = MessageQueue::MessageId(messageType, privateId);
    pthread_mutex_lock(&lock);
    MessageCallback *callback = messageCallbacks.find(id);
    if (callback != 0) {
      delete callback;
      messageCallbacks.erase(id);
    }
    pthread_mutex_unlock(&lock);
  }
  void MessageQueue::RemoveCallback(std::string& messageType) {
    RemoveCallback(messageType, "");
  }

  Message *MessageQueue::Receive(std::string& messageType,
				 std::string& privateId) {
    std::string id = MessageQueue::MessageId(messageType, privateId);
    while (true) {
      pthread_mutex_lock(&lock);
      MsqQueue *msgQueue = messageQueues.find(id);
      if ((msgQueue != NULL) && (!msgQueue.empty())) {
	Message *reply = msgQueue.pop();
	pthread_mutex_unlock(&lock);
	if (reply.Error() != 0) {
	  pthread_mutex_unlock(&lock);
	  throw Exception(reply.Error(), reply.ErrorDescription);
	}
	pthread_mutex_unlock(&lock);
	return reply.Clean();
      }
      // hmmm... wait for a new message on the queue and check that
      pthread_cond_wait(&go, &lock);
      // implicit unlock()
    }
  }
  Message *MessageQueue::Receive(std::string& messageType) {
    return Recieve(messageType, "");
  }

  void MessageQueue::Send(std::string& message) {
    try {
      deapi.Send(message);
    } catch (...) {
      throw Exception(ERR_SYSERR, "Unable to send message to Orc");
    }
  }
  void MessageQueue::Send(std::string& messageType, std::string& message) {
    Send(messageType, "", message);
  }
  void MessageQueue::Send(std::string& messageType, std::string& privateId,
			  std::string& message) {
    std::string wholeMessage = "{message_info={message_type=";
    wholeMessage += messageType;
    if (!privateId.empty()) {
      wholeMessage += "|private_id=";
      wholeMessage += privateId;
    }
    wholeMessage += "}";
    wholeMessage += message + "}";
    Send(message);
  }
      
  MessageQueue::~MessageQueue() {
    Logout();
  }

}

#endif

/*
  $log$
*/

File: util/deapi/java/main/dealingtools/orc/deapi/MessageQueue.java
/* ----------------------------------------------------------------------
   MessageQueue - makes asynchronous deapi look synchronous

   Original Author: John McMullan (2004)

   $Author: john $
   $Date: 2004/05/10 14:06:02 $
   $Source: /usr/export/cvsroot/util/deapi/java/main/dealingtools/orc/deapi/MessageQueue.java,v $
   $Revision: 1.1.1.1 $
   

   ---------------------------------------------------------------------- */

package dealingtools.orc.deapi;

import java.util.*;
import java.io.*;

public class MessageQueue extends Thread {
    private HashMap messageQueues;
    private HashMap messageCallbacks;
    private Connection deapi;

    public MessageQueue(String hostname, String service, String loginId,
			String password, boolean debug)
	throws DeapiException {
	initializeConnection(hostname, service);
	login(loginId, password, debug);
    }

    public MessageQueue(String hostname, String service, String loginId,
			String password) throws DeapiException {
	initializeConnection(hostname, service);
	login(loginId, password, false);
    }

    private void initializeConnection(String hostname, String service)
    throws DeapiException {
	try {
	    deapi = new Connection(hostname, service);
	} catch (IOException e) {
	    throw new DeapiException(DeapiErrors.NO_ORC,
				     "The Orc API is not running on ["
				     + hostname + ":" + service + "]");
	}
	messageQueues = new HashMap();
	messageCallbacks = new HashMap();

	// kick off recieve thread
	setDaemon(true);
	start();
    }

    public void run() {
	Message reply;
	LinkedList mQueue;
	MessageCallback callback;
	String id;

	while (true) {
	    try {
		reply = deapi.recieve();
		id = messageId(reply.type(), reply.privateId());
	    } catch (java.net.SocketException se) {
		// someone has closed our socket... finish...
		return;
	    } catch (Exception e) {
		throw new RuntimeException(e);
	    }

	    synchronized(messageCallbacks) {
		callback = (MessageCallback) messageCallbacks.get(id);
	    }
	    if (callback != null) {
		try {
		    if (reply.error() != 0) {
			callback.onError(reply.error(),
					 reply.errorDescription());
		    } else {
			callback.onMessage(reply.clean());
		    }
		} catch (DeapiException de) {
		    callback.onError(DeapiErrors.SYSERR,
				     "Unable to get error from message");
		}
	    } else {
		synchronized(messageQueues) {
		    // not for us, put it on the message queue, it might
		    // be a synchronous message
		    if ((mQueue = (LinkedList) messageQueues.get(id))
			== null) {
			mQueue = new LinkedList();
			messageQueues.put(id, mQueue);
		    }
		    mQueue.addFirst(reply);
		    // inform the recieve method that there's new message
		    // in case it's blocking on that one
		    messageQueues.notify();
		}
	    }
	}
    }
    
    public void registerCallback(String messageType, String privateId,
				 MessageCallback callback)
	throws DeapiException {
	String id = messageId(messageType, privateId);
	synchronized(messageCallbacks) {
	    messageCallbacks.put(id, callback);
	}
    }
    public void registerCallback(String messageType, MessageCallback callback)
	throws DeapiException {
	String id = messageId(messageType, null);
	synchronized(messageCallbacks) {
	    messageCallbacks.put(id, callback);
	}
    }

    public void removeCallback(String messageType, String privateId)
	throws DeapiException {
	String id = messageId(messageType, privateId);
	synchronized(messageCallbacks) {
	    messageCallbacks.remove(id);
	}
    }
    public void removeCallback(String messageType)
	throws DeapiException {
	String id = messageId(messageType, null);
	synchronized(messageCallbacks) {
	    messageCallbacks.remove(id);
	}
    }

    public void logout() {
	try {
	    send("{message_info={message_type=logout}}");
	    deapi.close();
	} catch (DeapiException de) {
	    // I can't think why I need to worry...
	} catch (IOException e) {
	    // ditto
	}
    }

    public void login(String loginId, String password, boolean debug)
	throws DeapiException {
	StringBuffer loginmsg =	new
	    StringBuffer("{message_info={message_type=login}|login_id=");
	loginmsg.append(loginId);
	loginmsg.append("|password=");
	loginmsg.append(password);
	if (debug) {
	    loginmsg.append("|debug=on");
	}
	loginmsg.append("}");

	send(loginmsg.toString());
	recieve("login");
    }

    public Message recieve(String messageType) throws DeapiException {
	return recieve(messageType, null);
    }
    public Message recieve(String messageType, String privateId)
			   throws DeapiException {
	Message reply;
	LinkedList mQueue;
	String id = messageId(messageType, privateId);

	while (true) {
	    synchronized(messageQueues) {
		reply = null;
		// look in our message queues for one of these
		if ((mQueue = (LinkedList) messageQueues.get(id)) != null) {
		    if (!mQueue.isEmpty()) {
			reply = (Message) mQueue.removeLast();
			if (reply.error() != 0) {
			    throw new DeapiException(reply.error(),
						     reply.errorDescription());
			}
			return reply.clean();
		    }
		}
		// hmmm wait until someone puts another message on the
		// queues
		try {
		    messageQueues.wait();
		} catch (InterruptedException e) {
		    throw new RuntimeException(e);
		}
	    }
	}
    }

    public void send(String message) throws DeapiException {
	try {
	    deapi.send(message);
	} catch (IOException e) {
	    throw new DeapiException(DeapiErrors.NO_ORC,
				     "Our Orc connection has gone!");
	}
    }
    public void send(String messageType, String message)
	throws DeapiException {
	String privateId = null;
	send(messageType, privateId, message);
    }
    public void send(String messageType, String privateId, String message)
	throws DeapiException {
	StringBuffer msg = new StringBuffer("{message_info={message_type=");
	msg.append(messageType);
	if (privateId != null) {
	    msg.append("|private_id = ");
	    msg.append(privateId);
	}
	msg.append("}|");
	msg.append(message);
	msg.append("}");
	try {
	    deapi.send(msg.toString());
	} catch (IOException e) {
	    throw new DeapiException(DeapiErrors.NO_ORC,
				     "Our Orc connection has gone!");
	}
    }

    public static String messageId(String messageType, String privateId) {
	if ((privateId == null) || (privateId.length() == 0)) {
	    return messageType;
	}
	return messageType + ":" + privateId;
    }
}

/*
  $log$
*/









